Fix issues introduced by createElement() warning

gaearon opened the pull request:
This builds on top of #6268.

I fixed a few style nits and indirection in it, and later I realized it introduces a bug in `cloneElement()`. As I was tracing this bug, I found that it already existed in `createElement()` in 15.x (#6879).

So I fixed the bug for both cases, added more tests, and shuffled them around to sit in the corresponding files.

spicyj commented on src/isomorphic/classic/element/ReactElement.js:
```diff
@@ -138,14 +151,16 @@ ReactElement.createElement = function(type, config, children) {
-      key = config.key === undefined ? null : '' + config.key;
     }
+
+    if (isValidConfigRefOrKey(config, 'ref')) {
```
I think this may be slower (in prod mode). We should avoid passing around object keys as strings whenever possible because engines can't optimize it as well.

spicyj replied:
Oh you fixed this.

spicyj commented on src/isomorphic/classic/element/__tests__/ReactElement-test.js:
```diff
@@ -170,6 +170,86 @@ describe('ReactElement', function() {
+
+    var element = React.createFactory(ComponentClass)(props);
+    expect(element.type).toBe(ComponentClass);
+    expect(element.key).toBe(null);
```
So in prod mode these would have different behavior, right? :(

spicyj replied:
(Not new, of course.)

gaearon replied:
Yea.

spicyj commented on src/isomorphic/classic/element/__tests__/ReactElement-test.js:
```diff
@@ -165,7 +165,59 @@ describe('ReactElement', function() {
     expect(element.ref).toBe('34');
-    var expectation = {foo:'56'};
+    var expectation = {foo: '56'};
+    Object.freeze(expectation);
```
What are the freezes for? (I guess they were here already too…)

gaearon replied:
I have no idea 😄 . They were there and spread like fire. I wouldn’t be surprised if the first freeze was added for a completely unrelated purpose. (Maybe from the pre-element times?)

spicyj commented:
Can you inline all the `var expectation =` now that they're only used once each? Otherwise looks great.

facebook-github-bot-9 commented:
@gaearon updated the pull request.

gaearon commented:
@zpao 

For the changelog:

* React: Fixed incorrect coercion of `ref` or `key` that are `undefined` in development (@gaearon in #6880)
* React: Fixed a false positive when passing other element’s props to `cloneElement` (@ericmatthys in #6268)
